src/vsr/replica.zig

```diff
@@ -1811,6 +1863,84 @@ pub fn ReplicaType(
             self.send_message_to_replica(message.header.replica, response);
         }
 
+        fn on_request_reply(self: *Self, message: *const Message) void {
+            assert(message.header.command == .request_reply);
+            assert(message.header.client != 0);
+
+            if (self.ignore_repair_message(message)) return;
+            assert(message.header.replica != self.replica);
+
+            const entry = self.client_sessions().get(message.header.client) orelse {
+                log.debug("{}: on_request_reply: ignoring, client not in table", .{self.replica});
+                return;
+            };
+            assert(entry.header.client == message.header.client);
+
+            if (entry.header.checksum != message.header.context) {
+                log.debug("{}: on_request_reply: ignoring, reply not in table " ++
+                    "(requested={} stored={})", .{
+                    self.replica,
+                    message.header.context,
+                    entry.header.checksum,
+                });
+                return;
+            }
+            assert(entry.header.size != @sizeOf(Header));
+            assert(entry.header.op == message.header.op);
+
+            const slot = self.client_sessions().get_slot_for_header(&entry.header).?;
+            if (self.client_replies.read_reply_sync(slot, entry) catch |err| {
+                assert(err == error.Busy);
```

Comment:
Hm, now that I look at this, shoudn't we get `error.Busy` from `read_reply` rather than `sync` version?

Reply:
Also, should probably log why we ignore here?

Reply:
I'll add a log. The caller must invoke `read_reply` immediately after `read_reply_sync` fails. The reason for handling the `Busy` case in `_sync` (and more generally, the reason for splitting out `_sync` at all) is to avoid needing `NextTick` in the read path. We would need an upper bound for `NextTick` contexts, and I don't think we can do better than `messages_max`.

Reply:
But I think Busy can be returned from either of the two functions? 

```diff
diff --git a/src/vsr/client_replies.zig b/src/vsr/client_replies.zig
index aac53c6e..fd82b59f 100644
--- a/src/vsr/client_replies.zig
+++ b/src/vsr/client_replies.zig
@@ -107,7 +107,7 @@ pub fn ClientRepliesType(comptime Storage: type) type {
             client_replies: *ClientReplies,
             slot: Slot,
             session: *const ClientSessions.Entry,
-        ) error{Busy}!?*Message {
+        ) ?*Message {
             const client = session.header.client;
 
             if (client_replies.writing.isSet(slot.index)) {
@@ -127,16 +127,6 @@ pub fn ClientRepliesType(comptime Storage: type) type {
                 return write_latest.?.message;
             }
 
-            if (client_replies.reads.available() == 0) {
-                log.debug("{}: read_reply: busy (client={} reply={})", .{
-                    client_replies.replica,
-                    session.header.client,
-                    session.header.checksum,
-                });
-
-                return error.Busy;
-            }
-
             return null;
         }
 
@@ -148,8 +138,18 @@ pub fn ClientRepliesType(comptime Storage: type) type {
             session: *const ClientSessions.Entry,
             callback: fn (*ClientReplies, *const vsr.Header, ?*Message, ?u8) void,
             destination_replica: ?u8,
-        ) void {
-            assert(client_replies.read_reply_sync(slot, session) catch unreachable == null);
+        ) error{Busy}!void {
+            assert(client_replies.read_reply_sync(slot, session) == null);
+
+            if (client_replies.reads.available() == 0) {
+                log.debug("{}: read_reply: busy (client={} reply={})", .{
+                    client_replies.replica,
+                    session.header.client,
+                    session.header.checksum,
+                });
+
+                return error.Busy;
+            }
 
             const read = client_replies.reads.acquire().?;
 
diff --git a/src/vsr/replica.zig b/src/vsr/replica.zig
index 079bf84e..99b6f6b9 100644
--- a/src/vsr/replica.zig
+++ b/src/vsr/replica.zig
@@ -1889,14 +1889,7 @@ pub fn ReplicaType(
             assert(entry.header.op == message.header.op);
 
             const slot = self.client_sessions().get_slot_for_header(&entry.header).?;
-            if (self.client_replies.read_reply_sync(slot, entry) catch |err| {
-                assert(err == error.Busy);
-
-                log.debug("{}: on_request_reply: ignoring, client_replies busy", .{
-                    self.replica,
-                });
-                return;
-            }) |reply| {
+            if (self.client_replies.read_reply_sync(slot, entry)) |reply| {
                 on_request_reply_read_callback(
                     &self.client_replies,
                     &entry.header,
@@ -1909,7 +1902,14 @@ pub fn ReplicaType(
                     entry,
                     on_request_reply_read_callback,
                     message.header.replica,
-                );
+                ) catch |err| {
+                    assert(err == error.Busy);
+
+                    log.debug("{}: on_request_reply: ignoring, client_replies busy", .{
+                        self.replica,
+                    });
+                    return;
+                };
             }
         }
 
@@ -3625,14 +3625,7 @@ pub fn ReplicaType(
             }
 
             const slot = self.client_sessions().get_slot_for_client(message.header.client).?;
-            if (self.client_replies.read_reply_sync(slot, entry) catch |err| {
-                assert(err == error.Busy);
-
-                log.debug("{}: on_request: ignoring (client_replies busy)", .{
-                    self.replica,
-                });
-                return;
-            }) |reply| {
+            if (self.client_replies.read_reply_sync(slot, entry)) |reply| {
                 on_request_repeat_reply_callback(
                     &self.client_replies,
                     &entry.header,
@@ -3645,7 +3638,14 @@ pub fn ReplicaType(
                     entry,
                     on_request_repeat_reply_callback,
                     null,
-                );
+                ) catch |err| {
+                    assert(err == error.Busy);
+
+                    log.debug("{}: on_request: ignoring (client_replies busy)", .{
+                        self.replica,
+                    });
+                    return;
+                };
             }
         }
 
```

This version looks a bit more natural to me, as we get an error when we try to schedule async work, and there isn't any resources for that. 

I guess the drawback here is that the callback is now called either zero or once, and that could be confusing?

Reply:
The zero-or-once-callback was the reason I was avoiding it, but this is indeed cleaner — we can even remove the `return` from within the catch. I'll update it.